Error while checking build/O2/fullCI for 786ecba44b7f5a5d45d5b3412f4cd0d10086a3df at 2024-07-02 12:58:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/13263-slc8_x86-64/0/Detectors/ITSMFT/common/reconstruction/src/DecodingStat.cxx:38:26: error: statement should be inside braces [readability-braces-around-statements]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13263/786ecba44b7f5a5d45d5b3412f4cd0d10086a3df/build_O2_fullCI/pretty.html).

--- a/Detectors/ITSMFT/common/reconstruction/include/ITSMFTReconstruction/AlpideCoder.h
+++ b/Detectors/ITSMFT/common/reconstruction/include/ITSMFTReconstruction/AlpideCoder.h
@@ -477,9 +477,6 @@ class AlpideCoder
       ChipPixelData currentChip;
       int localID = lidGetter(ID);
       if (seenChips.count(ID)) {
-        if (seenChips[ID]->isErrorSet()) {
-          continue;
-        }
         currentChip = *seenChips[ID];
       }
       AlpideCoder encoder;
@@ -492,6 +489,13 @@ class AlpideCoder
       reconstructedData.fill(0x00,
                              buffer.getSize() - reconstructedData.getSize());
 
+    auto hexToString = [](uint8_t v) {
+      std::stringstream ss;
+      ss << "0x" << std::setfill('0') << std::setw(2) << std::hex
+         << std::uppercase << (0xFF & v);
+      return ss.str();
+    };
+
     auto reportError = [&](std::string message) {
       LOG(error) << "Error during decoder verification: " << message;
       LOG(debug) << "Raw Data:";
@@ -499,17 +503,13 @@ class AlpideCoder
       uint8_t dataC = 0;
       int index = 1;
       while (buffer.next(dataC)) {
-        LOG(debug) << index << ". 0x" << std::setfill('0') << std::setw(2)
-                   << std::hex << std::uppercase << (0xFF & dataC);
-        ++index;
+        LOG(debug) << index++ << ". " << hexToString(dataC);
       }
       LOG(debug) << "Reconstructed Data:";
       reconstructedData.rewind();
       index = 1;
       while (reconstructedData.next(dataC)) {
-        LOG(debug) << index << ". 0x" << std::setfill('0') << std::setw(2)
-                   << std::hex << std::uppercase << (0xFF & dataC);
-        ++index;
+        LOG(debug) << index++ << ". " << hexToString(dataC);
       }
     };
 
@@ -522,6 +522,9 @@ class AlpideCoder
     // 2. Bunch counter for frame: the reconstructed buffer does contain the
     //    corresponding words, but has a dummy value.
 
+    ChipPixelData* currentChip = nullptr;
+    if (seenChipIDs.size())
+      currentChip = seenChips[seenChipIDs[0]];
     buffer.rewind();
     while (true) {
       uint8_t dataRec = 0;
@@ -550,6 +553,12 @@ class AlpideCoder
       buffer.current(dataRaw);
       if (dataRaw == dataRec) {
         if (isChipHeaderOrEmpty(dataRaw)) {
+          uint16_t ID = cidGetter(dataRaw & MaskChipID);
+          if (seenChips.count(ID)) {
+            currentChip = seenChips[ID];
+          } else {
+            currentChip = nullptr;
+          }
           // If the data correspond to the CHIPHEADER or CHIPEMPTY data words,
           // skip the next byte that represent bunch counters.
           buffer.next(dataRaw);
@@ -566,40 +575,102 @@ class AlpideCoder
         buffer.next(dataRaw);
         continue;
       }
-      if ((dataRaw & ERROR_MASK) == ERROR_MASK) {
-        // Placement of error triggers is arbitrary and we cannot guarantee
-        // the decoder to have processed the rest of the stream.
-        return false;
-      }
 
-      if (isChipHeader(dataRaw)) {
-        // If we encounter a mismatch when the raw data is at the beginning
-        // of the next chip, it might indicate the fact that an error occured
-        // during the decoding of this chip. In this case, we cannot continue
-        // the byte stream verification after this point.
-        uint16_t ID = cidGetter(dataRaw & MaskChipID);
-        if (seenChips.count(ID) && seenChips[ID]->isErrorSet()) {
-          LOG(warning) << "Chip " << ID
-                       << " could not be verified due to set errors: "
-                       << seenChips[ID]->reportErrors();
-          // TODO: Instead of returning, we can skip the part of the buffer
-          // that describes this chip and continue with the verification of
-          // the next one.
+      VerifierMismatchResult res =
+        handleVerifierMismatch(buffer, reconstructedData, currentChip);
+      switch (res) {
+        case VerifierMismatchResult::RESOLVED:
+          LOG(debug) << "Mismatch " << hexToString(dataRaw) << " / "
+                     << hexToString(dataRec)
+                     << " was resolved, able to continue verification";
+          continue;
+        case VerifierMismatchResult::EXPECTED_MISMATCH:
+          LOG(debug) << "Mismatch " << hexToString(dataRaw) << " / "
+                     << hexToString(dataRec)
+                     << " was expected, aborting the verification";
           return true;
+        case VerifierMismatchResult::UNEXPECTED_MISMATCH: {
+          // If the read bytes is not related to the special cases, report
+          // error.
+          std::stringstream errorStream;
+          errorStream
+            << "Unexpected byte mismatch during decoder verification. "
+               "Expected: "
+            << hexToString(dataRaw)
+            << ", Reconstructed: " << hexToString(dataRec);
+          reportError(errorStream.str());
         }
       }
-
-      // If the read bytes is not related to the special cases, report error.
-      std::stringstream errorStream;
-      errorStream << "Unexpected byte mismatch during decoder verification. "
-                     "Expected: 0x"
-                  << std::hex << std::uppercase << std::setfill('0')
-                  << std::setw(2) << (0xFF & dataRaw) << ", Reconstructed: 0x"
-                  << std::setfill('0') << std::setw(2) << (0xFF & dataRec);
-      reportError(errorStream.str());
       return false;
     }
     return true;
+  }
+
+  enum VerifierMismatchResult {
+    UNEXPECTED_MISMATCH, // Genuine mismatch, stop verification
+    EXPECTED_MISMATCH,   // Mismatch expected, need to abort verification
+    RESOLVED             // Mismatch resolved, can continue
+  };
+
+  static VerifierMismatchResult handleVerifierMismatch(
+    PayLoadCont& buffer, PayLoadCont& reconstructedData,
+    ChipPixelData* currentChip)
+  {
+    VerifierMismatchResult res = VerifierMismatchResult::UNEXPECTED_MISMATCH;
+    uint8_t dataRec = 0;
+    uint8_t dataRaw = 0;
+    reconstructedData.current(dataRec);
+    buffer.current(dataRaw);
+    auto inner = [&](int errIdx) {
+      if (res != VerifierMismatchResult::UNEXPECTED_MISMATCH ||
+          dataRaw == dataRec) {
+        // The mismatch was resolved, no need to check the rest of the errors
+        return;
+      }
+      switch (errIdx) {
+        case ChipStat::BusyViolation:
+        case ChipStat::DataOverrun:
+        case ChipStat::Fatal:
+        case ChipStat::BusyOn:
+        case ChipStat::BusyOff:
+        case ChipStat::TruncatedChipEmpty:
+        case ChipStat::TruncatedChipHeader:
+        case ChipStat::TruncatedRegion:
+        case ChipStat::TruncatedLondData:
+        case ChipStat::WrongDataLongPattern:
+        case ChipStat::NoDataFound:
+        case ChipStat::UnknownWord:
+        case ChipStat::RepeatingPixel:
+        case ChipStat::WrongRow:
+        case ChipStat::APE_STRIP_START:
+        case ChipStat::APE_ILLEGAL_CHIPID:
+        case ChipStat::APE_DET_TIMEOUT:
+        case ChipStat::APE_OOT:
+        case ChipStat::APE_PROTOCOL_ERROR:
+        case ChipStat::APE_LANE_FIFO_OVERFLOW_ERROR:
+        case ChipStat::APE_FSM_ERROR:
+        case ChipStat::APE_PENDING_DETECTOR_EVENT_LIMIT:
+        case ChipStat::APE_PENDING_LANE_EVENT_LIMIT:
+        case ChipStat::APE_O2N_ERROR:
+        case ChipStat::APE_RATE_MISSING_TRG_ERROR:
+        case ChipStat::APE_PE_DATA_MISSING:
+        case ChipStat::APE_OOT_DATA_MISSING:
+        case ChipStat::WrongDColOrder:
+        case ChipStat::InterleavedChipData:
+        case ChipStat::TruncatedBuffer:
+        case ChipStat::TrailerAfterHeader:
+        case ChipStat::FlushedIncomplete:
+        case ChipStat::StrobeExtended:
+        case ChipStat::WrongAlpideChipID:
+          break;
+        default:
+          LOG(error) << "Unknown error set by chip during verifier mismatch";
+      }
+    };
+    if (currentChip) {
+      currentChip->forEachSetError(inner);
+    }
+    return res;
   }
 
   static bool isChipEmpty(uint8_t v) { return (v & (~MaskChipID)) == CHIPEMPTY; }

